src/main/java/org/folio/service/processing/ParallelFileChunkingProcessor.java

```diff
@@ -116,33 +112,41 @@ public void process(JsonObject jsonRequest, JsonObject jsonParams) {
    * @param fileDefinition     fileDefinition entity
    * @param fileStorageService service to obtain file
    * @param params             parameters necessary for connection to the OKAPI
-   * @return Future
+   * @return Future parametrized by the total chunk counter value
    */
-  private Future<Void> processFile(FileDefinition fileDefinition, FileStorageService
-    fileStorageService, OkapiConnectionParams params) { // NOSONAR
-    Future<Void> future = Future.future();
-    SourceReader reader = new FileSystemStubSourceReader(this.vertx.fileSystem());
-    while (reader.hasNext()) {
-      reader.readNext().setHandler(readRecordsAr -> {
-        if (readRecordsAr.failed()) {
-          LOGGER.error("Can not read next chunk of records for the file {}", fileDefinition.getSourcePath());
-          future.fail(readRecordsAr.cause());
-        } else {
-          RawRecordsDto chunk = readRecordsAr.result();
-          postRawRecords(fileDefinition.getJobExecutionId(), chunk, params).setHandler(postedRecordsAr -> {
-            if (postedRecordsAr.failed()) {
-              future.fail(postedRecordsAr.cause());
-            } else {
-              if (!reader.hasNext()) {
-                LOGGER.info("All the chunks for file {} successfully sent.", fileDefinition.getSourcePath());
-                future.complete();
-              }
+  private Future<Integer> processFile(FileDefinition fileDefinition,
+                                      FileStorageService fileStorageService,
+                                      OkapiConnectionParams params) {
+    int totalChunkCounter = 0;
+    try {
+      SourceReader reader = new LocalStorageSourceReader(fileStorageService.getFile(fileDefinition.getSourcePath()));
+
+      /*
+        If one of the dedicated handlers for sending chunks is failed, then all the other senders have to be aware of that
+        in a terms of the target file processing.
+      */
+      AtomicBoolean canSendNextChunk = new AtomicBoolean(true);
+      for (List<String> records = reader.readNext(); records.size() > 0; records = reader.readNext(), totalChunkCounter++) {
+        /* Sending the next read chunk */
+        if (canSendNextChunk.get()) {
+          RawRecordsDto chunk = new RawRecordsDto().withLast(false).withRecords(records).withTotal(totalChunkCounter);
```

Comment:
Probably we should update RawRecordsDto data model - total is not really a total number of records for all the chunks except the last one. Definitely, the description of 'total' field should be updated and maybe the field name itself, e.g. 'counter', what do you think?